### PR TITLE
[sftp] Restrict sshfs perms to mapped ids

### LINKED ISSUE #3179: Mild security bug? Some writes in multipass to sshfs filesystem can have any userid the VM specifies.
Host: MacOS 13.5, ARM
VM: "Primary", Ubuntu 22.04
Multipass version info: 1.12.2+mac

My "Primary" multipass VM mounts `~/Shared` on my Mac rather than `~` because giving a VM access to my entire home directory feels unclean.  Anyway, `~/Shared` is mounted as `~/Shared` in the VM:

```
Mounts:         /Users/uccaoke/Shared => /home/ubuntu/Shared
                    UID map: 501:default
                    GID map: 20:default
```

My userid on the Mac is `uccaoke` (`501`) and my userid on the Ubuntu VM is `ubuntu` (`1000`).  I was tidying up some files in `~/Shared` on my Mac last night and notices that some of them were owned by `1000`.

All the files that were like this were `.gz` files and this made an easy test case.

In Ubuntu:

```
ubuntu@primary:~/Shared$ id
uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm),20(dialout),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),119(netdev),120(lxd)
ubuntu@primary:~/Shared$ touch test
ubuntu@primary:~/Shared$ gzip -k test
```

And then on the host Mac:

```
 ~/Shared ☛ ls -lah test*
-rw-rw-r--  1 uccaoke  staff     0B 28 Jul 18:52 test
-rw-rw-r--  1 1000     staff    25B 28 Jul 18:52 test.gz
```

As you can see, reproducibly, writing files with `gzip` in the VM creates a file on the host with numeric owner with the UID of `1000`.

Next of course, I was interested to see if this also happened if I were `root` in a VM as that would be quite bad:

```
ubuntu@primary:~/Shared$ sudo su -
root@primary:~# cd ~ubuntu/Shared/
root@primary:/home/ubuntu/Shared# touch rtest
root@primary:/home/ubuntu/Shared# gzip -k rtest
```

On the host:
```
 ~/Shared ☛ ls -lah rtest*
-rw-r--r--  1 uccaoke  staff     0B 28 Jul 18:57 rtest
-rw-r--r--  1 1000     staff    26B 28 Jul 18:57 rtest.gz
```

That's good at least, it means it's picking up the UID of 1000 from somewhere else.

On the VM there's some definite weirdness going on:

```
root@primary:/home/ubuntu/Shared# id
uid=0(root) gid=0(root) groups=0(root)
root@primary:/home/ubuntu/Shared# ls -lah rtest*
-rw-r--r-- 1 ubuntu ubuntu  0 Jul 28 18:57 rtest
-rw-r--r-- 1 ubuntu ubuntu 26 Jul 28 18:57 rtest.gz
```

Oh dear, what happens if we chown the file to root inside the VM?

On the VM:

```
root@primary:/home/ubuntu/Shared# chown root rtest
```

On the Mac:

```
~/Shared ☛ ls -lah rtest*
-rw-r--r--  1 root  staff     0B 28 Jul 18:57 rtest
-rw-r--r--  1 1000  staff    26B 28 Jul 18:57 rtest.gz
```

This seems potentially quite bad?  I haven't tested things like setuid bits but...

### PR DESCRIPTION
**Pull request by sharder996**
File operations in an instance in an sshfs mount are now restricted to only files that have a uid/gid mapping for them (host-to-guest).

Fixes #3179 

Public side of [#571](https://github.com/canonical/multipass-private/pull/571)

### CODE REVIEW

File: src/sshfs_mount/sftp_server.cpp [hunk 1]
```diff
@@ -493,17 +540,16 @@ int mp::SftpServer::handle_mkdir(sftp_client_message msg)
         return reply_failure(msg);
     }
 
-    QFileInfo current_dir(filename);
-    QFileInfo parent_dir(current_dir.path());
-    int rev_uid = reverse_uid_for(msg->attr->uid, parent_dir.ownerId());
-    int rev_gid = reverse_gid_for(msg->attr->gid, parent_dir.groupId());
+    int rev_uid = reverse_uid_for(parent_dir.ownerId(), parent_dir.ownerId());
+    int rev_gid = reverse_gid_for(parent_dir.groupId(), parent_dir.groupId());
```
Comment (sharder996) [inline src/sshfs_mount/sftp_server.cpp:544]: This was originally:
```
int rev_uid = reverse_uid_for(msg->attr->uid, parent_dir.ownerId());
int rev_gid = reverse_gid_for(msg->attr->gid, parent_dir.groupId());
```
but, libssh doesn't populate `msg->attr->uid` or the gid. Is there some scenario that I'm not thinking of where this does get populated? The same was also being done in `handle_open()`...

File: src/sshfs_mount/sshfs_mount.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ auto get_sshfs_exec_and_options(mp::SSHSession& session)
         }
         else
         {
-            sshfs_exec += " -o dcache_timeout=3";
+            sshfs_exec += " -o dcache_timeout=3 -o dir_cache=no";
```
[the remark this answers is not in the document]
Comment (sharder996) [reply src/sshfs_mount/sshfs_mount.cpp]: Hmmm, they both refer to "*directory cache*", so I guess `dcache_timeout` is obsolete if `dir_cache` is disabled.

**Comment by andrei-toterman** [comment]
Hey, @sharder996! I have functionally tested this and I have a few questions. Based on what I expected to happen, I think these changes are too restrictive, to the point in which I basically can't interact with anything that I do not own through a mapping, even if technically I should still have permissions to do so. But maybe I am not understanding the scope of this problem and these things are expected to happen.
For example:
- let's say i have a directory `dir` that I own on the host and inside `dir` there is `file` that I do not own but that has 0777 permissions i.e.
```bash
❯ ls -ld dir && ls -l dir 
drwxr-xr-x 2 andrei andrei 4096 Dec 11 15:17 dir
total 4
-rwxrwxrwx 1 1234 1234 5 Dec 11 15:17 file
```
if i do `mp mount dir instance:dir`, ownership is mapped as expected i.e. `dir` is owned by `ubuntu:ubuntu` and `file` is still owned by some non-existent `1234:1234` and still has 0777 permissions. So I would expect to be able to read from and write to this file even if `ubuntu:ubuntu` has no ownership over it, given that the file has R/W permissions for everyone.
- let's say again that i a have `dir` that I own on the host and inside it there is `file` that is owned by `1000:1000` (or whatever are the uid and gid of `ubuntu:ubuntu` inside `insatnce`). And I mount `dir` but without mapping `-u 1000:1000` and `-g 1000:1000` i.e. `mp mount dir instance:dir -u 1234:1234 -g 1234:1234`. Now, inside the instance the file appears as owed by `ubuntu:ubuntu` simply because there was no map necessary since it was owned on the host by the uid and gid of `ubuntu:ubuntu`
```
ubuntu@commending-bird:~/dir$ ls -l file
-rw-r--r-- 1 ubuntu ubuntu 14 Dec 11 15:04 file
```
So I would expect to be able to read/write from/to `file` but I get `Permission denied` since I did not explicitly provide a redundant mapping.
- similar to point 1, if I mount a directory inside an instance but without mapping the directory's uid and gid, I cannot `ls` that directory inside the instance even if the directory has 0777 permissions or mapping is not necessary because the uid and gid are already correct.
- if i map only the uid or only the gid of the owner of a file/directory to the uid/gid of `ubuntu`, I still cannot interact with the file/directory even if the permissions of the file shouldn't require me to own the file or be in the owning group in order to do so.

In summary, I feel like ownership and permissions do not play along as expected and that ownership is not honored if a redundant mapping if not provided for it. Please do let me know if these things are desired and if I am expecting the wrong things. Thanks!

**Review by andrei-toterman — APPROVED** [review]
Alright, let's get this in!
